PQAnalysis/io/traj_file/trajectory_reader.py

```diff
@@ -107,57 +120,271 @@ def read(self, topology: Topology | None = None) -> Trajectory:
+    ) -> Generator[AtomicSystem]:
         """
         A generator that yields the frames of the trajectory.
 
         The difference to the read method is that the read method returns the whole
-        trajectory at once, while the frame_generator yields the frames one after 
-        another. This is useful if the trajectory is very large and cannot be stored in memory.
+        trajectory at once, while the frame_generator yields the frames one after
+        another. This is useful if the trajectory is very large and cannot be stored
+        in memory.
 
-        This method is used to read the trajectory from the file. It reads the 
-        trajectory from the file and concatenates the lines of the same frame. 
+        This method is used to read the trajectory from the file. It reads the
+        trajectory from the file and concatenates the lines of the same frame.
         The frame information is then read from the concatenated string with the
         FrameReader class and a Frame object is created.
 
+        Parameters
+        ----------
+        trajectory_start : int, optional
+            The start index of the trajectory, by default 0.
+        trajectory_stop : int | None, optional
+            The last index of the trajectory, by default None, which then
+            set to the length of the trajectory.
+
+        Exceptions
+        ----------
+        IndexError
+            If trajectory_start is less than 0 or greater than the length
+            of the trajectory.
+            If trajectory_stop is less than 0 or greater than the length
+            of the trajectory.
+
         Yields
         ------
         Generator[AtomicSystem]
             The frames of the trajectory.
         """
+
+        if trajectory_stop is None:
+            trajectory_stop = self.length_of_traj
+
+        # If trajectory_start is less than 0 or greater than the
+        # length of the trajectory, raise an IndexError
+        if trajectory_start < 0 or trajectory_start > self.length_of_traj:
+            self.logger.error(
+                "start index is less than 0 or greater than the length of the trajectory",
+                exception=IndexError,
+            )
+
+        # If trajectory_stop is less than 0 or greater than the
+        # length of the trajectory, raise an IndexError
+        if trajectory_stop < 0 or trajectory_stop > self.length_of_traj:
+            self.logger.error(
+                "stop index is less than 0 or greater than the length of the trajectory",
+                exception=IndexError,
+            )
+
+        # If trajectory_start is greater than or equal to trajectory_stop,
+        # raise an IndexError
+        if trajectory_start >= trajectory_stop:
+            self.logger.error(
+                "start index is greater than or equal to the stop index",
+                exception=IndexError,
+            )
+
+        # Track the number of frames that have been yielded
```

Comment:
isn't this the number of frames that have been read and not that have been yielded?

Reply:
True. It was different in the initial implementation.